ipq40xx: RUTX10 wi-fi calibration fix

MinistrasP opened the pull request:
ipq40xx: RUTX10 wi-fi calibration fix

RUXT10 doesn't get proper (pre-)calibration data from nvmem-cells in device-tree.
This patch returns (pre-)calibration to 11-ath10k-caldata file.

Signed-off-by: Kasparas Elzbutas <elzkas@gmail.com>

Ansuel commented:
@MinistrasP can you give us more info about this?
Also think you should remove / comment the relevant node in the dts to handle precal in the script

MinistrasP commented:
With nvmem-cells calibration Wi-Fi doesn't work and i get this in logs:

`[   26.624919] ath10k_ahb a000000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   26.624975] ath10k_ahb a000000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   26.636860] ath10k_ahb a000000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   26.665258] ath10k_ahb a000000.wifi: Loading BDF type 0
[   26.670526] ath10k_ahb a000000.wifi: failed to fetch board data for bus=ahb,vendor=0000,device=0000,subsystem-vendor=0000,subsystem-device=0000 from ath10k/QCA4019/hw1.0/board-2.bin
[   26.890855] ath10k_ahb a000000.wifi: failed to fetch board-2.bin or board.bin from ath10k/QCA4019/hw1.0
[   26.890936] ath10k_ahb a000000.wifi: failed to fetch board file: -12
[   26.899288] ath10k_ahb a000000.wifi: could not probe fw (-12)
[   27.857667] ath10k_ahb a800000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   27.857723] ath10k_ahb a800000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   27.879159] ath10k_ahb a800000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   27.902151] ath10k_ahb a800000.wifi: Loading BDF type 0
[   27.907091] ath10k_ahb a800000.wifi: failed to fetch board data for bus=ahb,vendor=0000,device=0000,subsystem-vendor=0000,subsystem-device=0000 from ath10k/QCA4019/hw1.0/board-2.bin
[   28.131160] ath10k_ahb a800000.wifi: failed to fetch board-2.bin or board.bin from ath10k/QCA4019/hw1.0
[   28.131235] ath10k_ahb a800000.wifi: failed to fetch board file: -12
[   28.139576] ath10k_ahb a800000.wifi: could not probe fw (-12)
`

While with older 11-ath10k-caldata script, i get correct calibration:

`[   23.605592] ath10k_ahb a000000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   23.605648] ath10k_ahb a000000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   23.617533] ath10k_ahb a000000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   23.669392] ath10k_ahb a000000.wifi: Loading BDF type 0
[   23.670099] ath10k_ahb a000000.wifi: board_file api 2 bmi_id 0:16 crc32 3ebd5135
[   24.960759] ath10k_ahb a000000.wifi: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   24.960818] ath10k_ahb a000000.wifi: msdu-desc: 2500  skid: 32
[   24.997181] ath10k_ahb a000000.wifi: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   24.998117] ath10k_ahb a000000.wifi: wmi print 'free: 53252 iram: 13432 sram: 35752'
[   25.123863] ath10k_ahb a000000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[   26.605198] ath10k_ahb a800000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   26.605269] ath10k_ahb a800000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   26.618627] ath10k_ahb a800000.wifi: firmware ver 10.4b-ct-4019-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 6b2b5c5b
[   26.668936] ath10k_ahb a800000.wifi: Loading BDF type 0
[   26.669187] ath10k_ahb a800000.wifi: board_file api 2 bmi_id 0:17 crc32 3ebd5135
[   27.960743] ath10k_ahb a800000.wifi: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   27.960798] ath10k_ahb a800000.wifi: msdu-desc: 2500  skid: 32
[   27.996496] ath10k_ahb a800000.wifi: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   27.997488] ath10k_ahb a800000.wifi: wmi print 'free: 53252 iram: 13432 sram: 35752'
[   28.122034] ath10k_ahb a800000.wifi: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1`

This fix doesn't address the issue, on why it can't read the correct calibration data, but it makes the Wi-Fi work again.

Ansuel commented:
@MinistrasP can you post also the dts for the related device or a link?

MinistrasP commented:
DTS links:
https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts
https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi

Calibration configuration is stored in qcom-ipq4018-rutx.dtsi.

0xFelix commented:
AFAIR the bootloader mangles the device tree and that could be the issue why the nvmem-cells are missing after boot / do not work.

MinistrasP commented:
Yeah its probably like this:
https://github.com/openwrt/openwrt/commit/af425e42212d476dd95ec31f18d85b19004c9268

Ansuel commented:
@MinistrasP can you quickly check that and add some comments in the commit description and dts? pls

Also wth the bootloader does this.....

robimarko commented:
Can you fetch the live DTS from a running device using DTC?

I assume that bootloader is patching in broken partition table, however it can be fooled with a dummy disabled partitions node

Ansuel commented:
@robimarko this is the same trick with ax3600 right?

robimarko commented:
Yeah, if the newer 2016 U-boot is used

MinistrasP commented:
I'll do it on weekend or monday.

MinistrasP committed:
ipq40xx: RUTX10 Wi-Fi calibration fix

This fix for RUTX10 moves Wi-Fi calibration back from device tree (newer type) to 11-ath10k-caldata file (older).

U-Boot on these devices mangles the device tree, so nvmem-cell type calibration doesn't work.

Signed-off-by: Kasparas Elzbutas <elzkas@gmail.com>

MinistrasP commented:
> Can you fetch the live DTS from a running device using DTC?

Any idea on how to do that?

Ansuel commented:
/proc device-tree and you should navigate till you find the partitions node

Il Lun 18 Lug 2022, 09:42 Kasparas Elzbutas ***@***.***> ha
scritto:

> Can you fetch the live DTS from a running device using DTC?
>
> Any idea on how to do that?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/10282#issuecomment-1186873209>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQUAQ3VIRUNBCV6IFHTVUUDHRANCNFSM53U2ALQA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

MinistrasP commented:
Fetched DTS:

```
<stdout>: Warning (clocks_property): /soc/wifi@a800000:clocks: cell 0 is not a phandle reference
<stdout>: Warning (clocks_property): /soc/wifi@a800000:clocks: cell 2 is not a phandle reference
<stdout>: Warning (clocks_property): /soc/wifi@a800000:clocks: cell 4 is not a phandle reference

<stdout>: Warning (clocks_property): /soc/wifi@a000000:clocks: cell 0 is not a phandle reference
<stdout>: Warning (clocks_property): /soc/wifi@a000000:clocks: cell 2 is not a phandle reference
<stdout>: Warning (clocks_property): /soc/wifi@a000000:clocks: cell 4 is not a phandle reference

<stdout>: Warning (resets_property): /soc/wifi@a800000:resets: cell 0 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a800000:resets: cell 2 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a800000:resets: cell 4 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a800000:resets: cell 6 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a800000:resets: cell 8 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a800000:resets: cell 10 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a000000:resets: cell 0 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a000000:resets: cell 2 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a000000:resets: cell 4 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a000000:resets: cell 6 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a000000:resets: cell 8 is not a phandle reference
<stdout>: Warning (resets_property): /soc/wifi@a000000:resets: cell 10 is not a phandle reference

wifi@a800000 {
			compatible = "qcom,ipq4019-wifi";
			mac-address-increment = <0x03>;
			clocks = <0x05 0x3e 0x05 0x3f 0x05 0x40>;
			resets = <0x05 0x06 0x05 0x07 0x05 0x08 0x05 0x09 0x05 0x0a 0x05 0x0b>;
			nvmem-cells = <0x18 0x17>;
			clock-names = "wifi_wcss_cmd\0wifi_wcss_ref\0wifi_wcss_rtc";
			status = "ok";
			interrupts = <0x00 0x30 0x01 0x00 0x31 0x01 0x00 0x32 0x01 0x00 0x33 0x01 0x00 0x34 0x01 0x00 0x35 0x01 0x00 0x36 0x01 0x00 0x37 0x01 0x00 0x38 0x01 0x00 0x39 0x01 0x00 0x3a 0x01 0x00 0x3b 0x01 0x00 0x3c 0x01 0x00 0x3d 0x01 0x00 0x3e 0x01 0x00 0x3f 0x01 0x00 0xa9 0x04>;
			reg = <0xa800000 0x200000>;
			reset-names = "wifi_cpu_init\0wifi_radio_srif\0wifi_radio_warm\0wifi_radio_cold\0wifi_core_warm\0wifi_core_cold";
			interrupt-names = "msi0\0msi1\0msi2\0msi3\0msi4\0msi5\0msi6\0msi7\0msi8\0msi9\0msi10\0msi11\0msi12\0msi13\0msi14\0msi15\0legacy";
			nvmem-cell-names = "pre-calibration\0mac-address";
		};

edma@c080000 {
			qcom,page-mode = <0x00>;
			compatible = "qcom,ess-edma";
			qcom,num_gmac = <0x02>;
			qcom,poll_required = <0x01>;
			qcom,rx_head_buf_size = <0x604>;
			status = "okay";
			interrupts = <0x00 0x41 0x01 0x00 0x42 0x01 0x00 0x43 0x01 0x00 0x44 0x01 0x00 0x45 0x01 0x00 0x46 0x01 0x00 0x47 0x01 0x00 0x48 0x01 0x00 0x49 0x01 0x00 0x4a 0x01 0x00 0x4b 0x01 0x00 0x4c 0x01 0x00 0x4d 0x01 0x00 0x4e 0x01 0x00 0x4f 0x01 0x00 0x50 0x01 0x00 0xf0 0x01 0x00 0xf1 0x01 0x00 0xf2 0x01 0x00 0xf3 0x01 0x00 0xf4 0x01 0x00 0xf5 0x01 0x00 0xf6 0x01 0x00 0xf7 0x01 0x00 0xf8 0x01 0x00 0xf9 0x01 0x00 0xfa 0x01 0x00 0xfb 0x01 0x00 0xfc 0x01 0x00 0xfd 0x01 0x00 0xfe 0x01 0x00 0xff 0x01>;
			qcom,mdio_supported;
			reg = <0xc080000 0x8000>;

			gmac0 {
				mac-address = [00 1e 42 4a 4c 27];
				local-mac-address = [00 1e 42 4a 4c 27];
				vlan_tag = <0x01 0x1f>;
			};

			gmac1 {
				qcom,forced_speed = <0x3e8>;
				mac-address = [00 1e 42 4a 4c 28];
				local-mac-address = [00 1e 42 4a 4c 28];
				qcom,poll_required = <0x01>;
				qcom,phy_mdio_addr = <0x04>;
				qcom,forced_duplex = <0x01>;
				vlan_tag = <0x02 0x20>;
			};
		};

wifi@a000000 {
			compatible = "qcom,ipq4019-wifi";
			mac-address-increment = <0x02>;
			clocks = <0x05 0x3b 0x05 0x3c 0x05 0x3d>;
			resets = <0x05 0x00 0x05 0x01 0x05 0x02 0x05 0x03 0x05 0x04 0x05 0x05>;
			nvmem-cells = <0x16 0x17>;
			clock-names = "wifi_wcss_cmd\0wifi_wcss_ref\0wifi_wcss_rtc";
			status = "ok";
			interrupts = <0x00 0x20 0x01 0x00 0x21 0x01 0x00 0x22 0x01 0x00 0x23 0x01 0x00 0x24 0x01 0x00 0x25 0x01 0x00 0x26 0x01 0x00 0x27 0x01 0x00 0x28 0x01 0x00 0x29 0x01 0x00 0x2a 0x01 0x00 0x2b 0x01 0x00 0x2c 0x01 0x00 0x2d 0x01 0x00 0x2e 0x01 0x00 0x2f 0x01 0x00 0xa8 0x04>;
			reg = <0xa000000 0x200000>;
			reset-names = "wifi_cpu_init\0wifi_radio_srif\0wifi_radio_warm\0wifi_radio_cold\0wifi_core_warm\0wifi_core_cold";
			interrupt-names = "msi0\0msi1\0msi2\0msi3\0msi4\0msi5\0msi6\0msi7\0msi8\0msi9\0msi10\0msi11\0msi12\0msi13\0msi14\0msi15\0legacy";
			nvmem-cell-names = "pre-calibration\0mac-address";
		};
```

MinistrasP commented:
Partitions:

```
xt25f128b@0 {
				compatible = "jedec,spi-nor\0n25q128a11";
				sector-size = <0x10000>;
				density = <0x1000000>;
				#address-cells = <0x01>;
				#size-cells = <0x01>;
				reg = <0x00>;
				spi-max-frequency = <0x16e3600>;

				partitions {
					compatible = "fixed-partitions";
					#address-cells = <0x01>;
					#size-cells = <0x01>;

					partition@190000 {
						label = "0:APPSBLENV";
						reg = <0x190000 0x10000>;
					};

					partition@0 {
						read-only;
						label = "0:SBL1";
						reg = <0x00 0x40000>;
					};

					partition@a0000 {
						read-only;
						label = "0:QSEE";
						reg = <0xa0000 0x60000>;
					};

					partition@160000 {
						read-only;
						label = "0:CDT";
						reg = <0x160000 0x10000>;
					};

					partition@2e0000 {
						label = "0:ART";
						reg = <0x2e0000 0x10000>;
					};

					partition@40000 {
						read-only;
						label = "0:MIBIB";
						reg = <0x40000 0x20000>;
					};

					partition@80000 {
						label = "0:BOOTCONFIG1";
						reg = <0x80000 0x20000>;
					};

					partition@300000 {
						label = "0:CONFIG_RW";
						reg = <0x300000 0x10000>;
					};

					partition@1a0000 {
						read-only;
						label = "0:APPSBL";
						reg = <0x1a0000 0xa0000>;
					};

					partition@170000 {
						label = "0:CDT_1";
						reg = <0x170000 0x10000>;
					};

					partition@240000 {
						label = "0:APPSBL_1";
						reg = <0x240000 0xa0000>;
					};

					partition@2f0000 {
						read-only;
						label = "0:CONFIG";
						reg = <0x2f0000 0x10000>;
					};

					partition@310000 {
						label = "0:EVENTSLOG";
						reg = <0x310000 0x90000>;
					};

					partition@100000 {
						label = "0:QSEE_1";
						reg = <0x100000 0x60000>;
					};

					partition@180000 {
						label = "0:DDRPARAMS";
						reg = <0x180000 0x10000>;
					};

					partition@60000 {
						label = "0:BOOTCONFIG";
						reg = <0x60000 0x20000>;
					};
				};
			};
```

MinistrasP commented:
Hmm, shouldn't there be these nodes
```
macaddr_config_0: macaddr@0 {
					reg = <0x0 0x6>;
				};
```

```
precal_art_1000: precal@1000 {
					reg = <0x1000 0x2f20>;
				};

				precal_art_5000: precal@5000 {
					reg = <0x5000 0x2f20>;
				};
```
Under partitions:
```
partition@2e0000
partition@2f0000
```
That's weird.

Leo-PL commented:
@robimarko @MinistrasP , there is an easier way:
`cat /sys/firmware/fdt`
This will show you the raw .dtb as passed from bootloader, then you can process it with DTC further.

chunkeey commented:
@MinistrasP, @Ansuel 
Should this be merged into openwrt-22.03 as is? So the release doesn't suffer from broken wifi on the RUTX10?

chunkeey commented:
Well, openwrt's 22.03 image for the RUTX10 should have a functioning wifi now...

Ansuel commented:
@chunkeey thanks for fixing this on openwrt 22.03.
@MinistrasP @robimarko to sum up this, uboot overwrite the partition and we can't fix this or we lose dual partition support. Correct?

Think there isn't a real solution than reverting the nvmem change... saddly.

robimarko commented:
Basically as far as I remember that sums it up

Ansuel commented:
@MinistrasP Can you rebase and add some comments about this in the DTS? Just to not encounter this problem again in the future if we have to make other magic?

chunkeey commented:
well, I've back(or forward?)ported the openwrt-22.03 change to master.
